src/consumer/mod.rs

```diff
@@ -12,8 +12,53 @@ const STATUS_RUNNING: usize = 0;
 const STATUS_QUIET: usize = 1;
 const STATUS_TERMINATING: usize = 2;
 
-type JobRunner<E> = dyn Fn(Job) -> Result<(), E> + Send + Sync;
-type BoxedJobRunner<E> = Box<JobRunner<E>>;
+/// Implementors of this trait can be registered to run jobs in a `Consumer`.
+///
+/// # Example
+///
+/// Create a worker with all default options, register a single handler (for the `foo` job
+/// type), connect to the Faktory server, and start accepting jobs.
+/// The handler is a struct that implements `JobRunner`.
+///
+/// ```no_run
+/// use faktory::{ConsumerBuilder, JobRunner, Job};
+/// use std::io;
+///
+/// struct MyHandler {
+///     config: String,
+/// }
+/// impl JobRunner<io::Error> for MyHandler {
+///    fn run(&self, job: Job) -> Result<(), io::Error> {
+///       println!("config: {}", self.config);
+///       println!("job: {:?}", job);
+///       Ok(())
+///   }
+/// }
+///
+/// let mut c = ConsumerBuilder::default();
+/// let handler = MyHandler {
+///    config: "bar".to_string(),
+/// };
+/// c.register("foo", handler);
+/// let mut c = c.connect(None).unwrap();
+/// if let Err(e) = c.run(&["default"]) {
+///     println!("worker failed: {}", e);
+/// }
+/// ```
+pub trait JobRunner<E>: Send + Sync {
+    /// A handler function that runs a job.
+    fn run(&self, job: Job) -> Result<(), E>;
+}
+// Implements JobRunner for a closure that takes a Job and returns a Result<(), E>
+impl<E, F> JobRunner<E> for F
```

Comment:
Unfortunately I think this won't do quite what you want — because this is a blanket implementation, no other implementations will be allowed, so no-one will be able to `impl JobRunner` themselves. It sucks, but it's a reality of the way the orphan rules in Rust works. I think the way we'll have to do this instead is to implement the trait just for `Box<dyn JobRunner>` (i.e., not as a blanket impl), leave `register` the way it currently is, and then add a `register_runner` method that takes `R: JobRunner`.

Reply:
Correct me if I am wrong, but my understanding of the orphan rules was that either the trait or the type for a given `impl` had to be in the crate.
You should be able to implement the trait for any type that is declared in the same crate as the implementation of the trait.
I would not expect that a user should be able to implement the `JobRunner` for a closure manually.

Reply:
This led me down quite a rabbit hole! It wasn't actually the orphan rules I was concerned about (mis-typed), it was the overlapping impls rules. Specifically, I was concerned that having a blanket impl here would prevent any downstream crate from implementing the trait since the impl would be potentially-conflicting with the blanket impl. But, it turns out, Rust doesn't consider this a hazard for `Fn` specifically because `Fn`'s set of implementations are "closed" — that is, the Rust standard library has essentially committed to not automatically implementing `Fn` for any new types. This is not the case for all traits — for instance, if you `s/Fn/Unpin/` (just as an example exercise), this blanket impl _does_ prevent implementing the trait for any other type, precisely because std reserves the right to `impl Unpin` automatically for user types in future versions. Interestingly, using a trait defined in the current crate as a bound is _also_ considered okay, because arriving at a conflict would either require that we add a blanket impl of that second trait (which is a breaking change), or the consuming crate would need to be the one to impl the second trait themselves, at which point they'll be told about the conflict.

All that to say, this is indeed okay, it just turns out to be for a surprisingly complex reason :sweat_smile: